shared/ssa/codeql/ssa/Ssa.qll

```diff
@@ -1163,4 +1163,480 @@ module Make<LocationSig Location, InputSig<Location> Input> {
+        DefinitionExt def, SourceVariable v, BasicBlock bb, int i, BasicBlock input
+      ) {
+        lastRefBeforeRedefExt(def, v, bb, i, input, this)
+      }
+    }
+
+    cached
+    private newtype TNode =
+      TParamNode(DfInput::Parameter p) { DfInput::ssaDefInitializesParam(_, p) } or
+      TExprNode(DfInput::Expr e, Boolean isPost) {
+        e = getARead(_)
+        or
+        DfInput::ssaDefAssigns(_, e) and
+        isPost = false
+      } or
+      TSsaDefinitionNode(DefinitionExt def) or
+      TSsaInputNode(SsaInputDefinitionExt def, BasicBlock input) {
+        def.hasInputFromBlock(_, _, _, _, input)
+      }
+
+    /**
+     * A data flow node that we need to reference in the value step relation.
+     *
+     * Note that only the `SsaNode` subclass is expected to be added as additional
+     * nodes in `DataFlow::Node`. The other subclasses are expected to already be
+     * present and are included here in order to reference them in the step relation.
+     */
+    abstract private class NodeImpl extends TNode {
+      /** Gets the location of this node. */
+      abstract Location getLocation();
+
+      /** Gets a textual representation of this node. */
+      abstract string toString();
+    }
+
+    final class Node = NodeImpl;
+
+    /** A parameter node. */
+    private class ParameterNodeImpl extends NodeImpl, TParamNode {
+      private DfInput::Parameter p;
+
+      ParameterNodeImpl() { this = TParamNode(p) }
+
+      /** Gets the underlying parameter. */
+      DfInput::Parameter getParameter() { result = p }
+
+      override string toString() { result = p.toString() }
+
+      override Location getLocation() { result = p.getLocation() }
+    }
+
+    final class ParameterNode = ParameterNodeImpl;
+
+    /** A (post-update) expression node. */
+    abstract private class ExprNodePreOrPostImpl extends NodeImpl, TExprNode {
+      DfInput::Expr e;
+      boolean isPost;
+
+      ExprNodePreOrPostImpl() { this = TExprNode(e, isPost) }
+
+      /** Gets the underlying expression. */
+      DfInput::Expr getExpr() { result = e }
+
+      override Location getLocation() {
+        exists(BasicBlock bb, int i |
+          e.hasCfgNode(bb, i) and
+          result = bb.getNode(i).getLocation()
+        )
+      }
+    }
+
+    final class ExprNodePreOrPost = ExprNodePreOrPostImpl;
+
+    /** An expression node. */
+    private class ExprNodeImpl extends ExprNodePreOrPostImpl {
+      ExprNodeImpl() { isPost = false }
+
+      override string toString() { result = e.toString() }
+    }
+
+    final class ExprNode = ExprNodeImpl;
+
+    /** A post-update expression node. */
+    private class ExprPostUpdateNodeImpl extends ExprNodePreOrPostImpl {
+      ExprPostUpdateNodeImpl() { isPost = true }
+
+      /** Gets the pre-update expression node. */
+      ExprNode getPreUpdateNode() { result = TExprNode(e, false) }
+
+      override string toString() { result = e.toString() + " [postupdate]" }
```

Comment:
I simply copied it from the [variable capture library](https://github.com/github/codeql/blob/main/shared/dataflow/codeql/dataflow/VariableCapture.qll#L763). However, these nodes should never actually be rendered, because they will be mapped to existing `DataFlow::Node`s with appropriate `toString`s.